### ISSUE TITLE
mem.replacementSize adjacent matches bug

### ISSUE THREAD

**Issue by jumpnbrownweasel** [ziglang/zig#8454]
Adjacent matches are not handled correctly by mem.replacementSize.

```zig
test "replacementSize" {
    std.testing.expectEqual(@as(usize, 2), std.mem.replacementSize(u8, "\\n\\n", "\\n", "\n"));
}
```
Result in: ```Test [1/42] test "replacementSize"... expected 2, found 3```

This was reported on discord by @cryptocode here:
https://discord.com/channels/605571803288698900/605572581046747136/829096724500840488

I'm putting together a PR and will be working with cryptocode on it.

**Comment by jumpnbrownweasel** [comment]
@cryptocode Here is a draft PR to review: #8455 

I added two tests for the 'replace' method as well as for 'replacementSize'. The changes to the existing tests for the 'replace' method were done because they were incorrect and just happened to work -- the size of two expected results just happened to be the same length.

I did not include any changes to address the O(n^2) problem that was pointed out on discord. The 'replace()' fn also has the same problem and I think it should be addressed separately with quite a bit more testing added.
